### ISSUE TITLE
job.properties fails validation if job not successful 

### ISSUE THREAD

**Issue by nonhermitian** [Qiskit/qiskit-ibmq-provider#116]
### Informations

Moved from: https://github.com/Qiskit/qiskit-terra/issues/2408

- **Qiskit version**: master
- **Python version**:
- **Operating system**:

### What is the current behavior?
```python
q = QuantumRegister(2)
c = ClassicalRegister(2)
qc = QuantumCircuit(q, c)

qc.h(q[0])
qc.cx(q[0], q[1])
qc.measure(q, c)

backend = IBMQ.get_backend('ibmqx4')
job = execute(qc, backend)
job.properties()
```
gives 
```python

ModelValidationError: {'gates': {0: {'qubits': ['Missing data for required field.'], 'parameters': {0: {'unit': ['Missing data for required field.'], 'date': ['Missing data for required field.'], 'name': ['Missing data for required field.'], 'value': ['Missing data for required field.']}}, 'gate': ['Missing data for required field.']}}, 'general': {0: {'unit': ['Missing data for required field.'], 'date': ['Missing data for required field.'], 'name': ['Missing data for required field.'], 'value': ['Missing data for required field.']}}, 'backend_name': ['Missing data for required field.'], 'backend_version': ['Missing data for required field.'], 'last_update_date': ['Missing data for required field.'], 'qubits': ['Missing data for required field.']}
```

if the job is not yet finished.

### What is the expected behavior?
I would expect the properties to be available as nothing in the properties object actually requires the job having been successful.  Indeed, it is nothing but the backend properties.

**Comment by diego-plan9** [comment]
> I would expect the properties to be available as nothing in the properties object actually requires the job having been successful. Indeed, it is nothing but the backend properties.

From the API side, it seems the properties are being fetched at the time the job is completed - mostly as an attempt to ensure that they are as up to date as possible (taking as a reference point the "time the job was actually executed").

There are also other corner cases that might prevent the properties being available at all times - for example, jobs that error out early before reaching the devices. We might be able to push the availability of `properties()` a bit earlier in some cases, but that is a question probably better suited to the API team - can you go through the cases and the desired behaviour with them?

On the provider side, a reasonable compromise might be to clarify in the documentation that the `properties()` of a job are not available in all cases (ideally with a bit more of info about which cases), and revise the function to handle empty `properties()` gracefully - which indeed is not happening at the moment.

**Comment by nonhermitian** [comment]
So if you look how the `properties` are being used, you see that things like transpiler optimizations are using the `properties` as they are when the job is transpiled or executed, not when the job is actually run.  So according to what you say, if the queue is very long, I could have transpiled for one set of properties, but then the job instance returns a different one.  I do not think that this is what would be expected.  This is why I would have the properties immediately tagged to the job.  This also would get around the issue with an errored job, as you attach the properties before hitting the backend.

**Comment by nonhermitian** [comment]
I will follow up with the API team.

**Comment by pacomf** [comment]
@nonhermitian @diego-plan9 The problem with the properties is that if for example you recovery the properties of the backend when you run a job... but the job is in a large queue and when the job is executed the properties are differents (you were in the middle of a calibration waiting in the queue), what is the correct behaviour?

**Comment by nonhermitian** [comment]
So, the properties are mainly used for transpiling (noise adapative transpiling), picking good initial layouts, etc.  This is all done using the properties as defined at the time the job was executed, not sometime later.  So I would attach the properties actually used in the circuit generation / transpilation, rather than a different set.

**Comment by pacomf** [comment]
ok! I understand... so we need to store in the Job the properties that the backend had when the job was created, not when the job was executed, right?

**Comment by nonhermitian** [comment]
That would be the desired behavior.  I believe the backend returns the properties to the API with each calibration, so they should be accessible.

**Comment by pacomf** [comment]
Yes, it is accesible, but we though that the right information to store in the job, is the properties that the backend had in the momento to run the job.

**Comment by diego-plan9** [comment]
Assigned @jyu00 (welcome to the team) to the part of the issue we can handle at the moment in the provider:

> and revise the function to handle empty properties() gracefully - which indeed is not happening at the moment.

Jessie, can you make sure that we catch the cases where the endpoint returns empty properties, and in that case make the `Job` function return `None` ? Ideally adding  a note in the properties that reflects the discussion on this issue regarding the properties not being available in all cases, that we will make a bit more specific once the discussion progresses.

**Comment by diego-plan9** [comment]
After #139 (thanks @jyu00 !) , hopefully the provider should be ready for dealing with empty properties gracefully. Leaving the issue open for discussion and keeping track of API updates about the exact point in time where a job's properties will be available.

**Comment by jyu00** [comment]
@nonhermitian The API now returns the backend properties at the time the job was created, if the job hasn't finished. Otherwise it returns the properties at the time the job is executed. If this is the desired behavior I'll update the `job.properties()` docstring and close this issue.

**Comment by nonhermitian** [comment]
I think this point deserves some discussion.  Currently we do not explicitly say whether the properties are returned at submission time or run time.  If we mix the two, the situation becomes even more confusing.  We may need to have both the submission properties, useful for transpilation, and the actual run time properties, useful for seeing what the device looked like when actually run.  

This discussion is probably wider than this issue.
